### DISCUSSION TITLE
Most efficient way to construct and use sparse matrix as variables?

### DISCUSSION THREAD

**Discussion by 1ucasvb** [cvxpy/cvxpy#2039] [Q&A]
I have a **massive** SDP which I managed to formulate in a very nice sparse way, but the pre-solving stage takes a very long time (over a day), but solving takes only a few hours. I believe it can be optimized significantly in CVXPY, but many of the techniques I attempted result in internal errors.

My variable is a PSD matrix `Y` (say, NxN), which exploiting sparsity, I can write in terms of small set of M scalar variables `y[i]`, and the positivity and other constraints in terms of other smaller matrices `Z_j` involving subsets of those variables.

I currently construct `Y` and `Z_j` directly as a sum of many terms of the form `(CVXPY scalar variable) * (a sparse matrix with a single 1)`, which creates very large parse trees internally. Nothing else has worked.

I originally tried to assemble the matrices using a vector variable `y` and auxiliary sparse (N² x M) matrices `A`, i.e., such that `Y = cp.reshape(A @ y, (N,N))`, but this does not work. The objective function or some other constraint usually gives an error regarding "lazy evaluation" and some feature not being implemented. It seems the vector structure of `y` breaks the sort of "pick and choose" strategy I'm using exploit the sparsity, that is, I'm using `y` as a list of numbers, not as an actual vector. This confuses CVXPY apparently.

I can get around this by first creating a vector `v` using scalar variables `y[i]` by summing over `i` and adding a sparse matrix of size `(N,1)` with a single 1 at position `i` multiplied by `y[i]`. Then the matrix trick in the last paragraph works, but I basically get the same large parse tree, as the sum in the definition of `v` is basically going to be stored symbolically as well, with thousands of terms.

Writing the matrices explicitly in terms of isolated scalar variables is the only thing that worked, and I also had to write the objective function manually, otherwise I get the same "lazy evaluation/not implemented" error then.

The result of this is that ultimately, even if everything is written in terms of sparse matrices and matrix multiplication and reshapings, the actual objects being manipulated are very large in CVXPY's representation, and mostly made of trivial sums of sparse matrices with single 1s at specific positions, just to "assemble" the sparse structure by hand.

Is there a better way of constructing such sparse matrices "by hand" in CVXPY, resulting in more efficient representations for the pre-solving stage? I just want to construct sparse matrices with entries as CVXPY scalar variables or expressions, instead of having to "add" them together one by one for each entry.

I have also tried using the `sparsity` option in declaring a matrix variable, but this seems to have no effect, and it's a completely undocumented feature. The solver still tries to allocate a massive dense matrix.

**Comment by SteveDiamond** [comment]
Can you share your code? Also what version of cvxpy are you using? Generally speaking if you have a sparse matrix variable X with nnz nonzeros and shape (m, n) it should be possible to express X as ``X = reshape(A @ Variable(nnz), (m, n))`` for some sparse matrix A.

**Comment by 1ucasvb** [comment]
Sure. I will try to post a minimum example in the coming days.
